nistats/first_level_model.py

```diff
@@ -0,0 +1,561 @@
+
+def run_glm(Y, X, noise_model='ar1', bins=100, n_jobs=1, verbose=0):
+    """ GLM fit for an fMRI data matrix
+
+    Parameters
+    ----------
+    Y : array of shape (n_time_points, n_voxels)
+        The fMRI data.
+
+    X : array of shape (n_time_points, n_regressors)
+        The design matrix.
+
+    noise_model : {'ar1', 'ols'}, optional
+        The temporal variance model. Defaults to 'ar1'.
+
+    bins : int, optional
+        Maximum number of discrete bins for the AR(1) coef histogram.
+
+    n_jobs : int, optional
+        The number of CPUs to use to do the computation. -1 means
+        'all CPUs'.
+
+    verbose : int, optional
+        The verbosity level. Defaut is 0
+
+    Returns
+    -------
+    labels : array of shape (n_voxels,),
+        A map of values on voxels used to identify the corresponding model.
+
+    results : dict,
+        Keys correspond to the different labels values
+        values are RegressionResults instances corresponding to the voxels.
+    """
+    acceptable_noise_models = ['ar1', 'ols']
+    if noise_model not in acceptable_noise_models:
+        raise ValueError(
+            "Acceptable noise models are {0}. You provided 'noise_model={1}'".\
+                format(acceptable_noise_models, noise_model))
+
+    if Y.shape[0] != X.shape[0]:
+        raise ValueError(
+            'The number of rows of Y should match the number of rows of X.'
+            ' You provided X with shape {0} and Y with shape {1}'.\
+                format(X.shape, Y.shape))
+
+    # Create the model
+    ols_result = OLSModel(X).fit(Y)
+
+    if noise_model == 'ar1':
+        # compute and discretize the AR1 coefs
+        ar1 = ((ols_result.resid[1:] * ols_result.resid[:-1]).sum(axis=0) /
+               (ols_result.resid ** 2).sum(axis=0))
+        del ols_result
+        ar1 = (ar1 * bins).astype(np.int) * 1. / bins
+        # Fit the AR model acccording to current AR(1) estimates
+        results = {}
+        labels = ar1
+        # Parallelize by creating a job per ARModel
+        vals = np.unique(ar1)
+        ar_result = Parallel(n_jobs=n_jobs, verbose=verbose)(
+            delayed(_ar_model_fit)(X, val, Y[:, labels == val]) for val in vals)
+        for val, result in zip(vals, ar_result):
+            results[val] = result
+        del vals
+        del ar_result
+
+    else:
+        labels = np.zeros(Y.shape[1])
+        results = {0.0: ols_result}
+
+    return labels, results
+
+
+class FirstLevelModel(BaseEstimator, TransformerMixin, CacheMixin):
+    """ Implementation of the General Linear Model for single session fMRI data
+
+    Parameters
+    ----------
+
+    t_r: float
+        This parameter indicates repetition times of the experimental runs.
+        It is necessary to correctly consider times in the design matrix.
+        This parameter is also passed to nilearn.signal.clean.
+        Please see the related documentation for details.
+
+    slice_time_ref: float, optional (default 0.)
```

Comment:
The problem with your approach is that you no longer support irregular sampling, that was mad possible by directly providing frame_times. Leaving the possibility to use frame_times would yield now 3 arguments instead of 1. I understand that it makes the general case simpler (e.g. using BIDS convention that provide a TR and default slice_time, the frame_times would be inferred automatically).